libnetwork/iptables/firewalld.go

```diff
@@ -232,14 +248,39 @@ func setupDockerZone() error {
 		target:      "ACCEPT",
 	}
 	if err := connection.sysConfObj.Call(dbusInterface+".config.addZone", 0, dockerZone, dz.settings()).Err; err != nil {
-		return err
-	}
-	// Reload for change to take effect
-	if err := connection.sysObj.Call(dbusInterface+".reload", 0).Err; err != nil {
-		return err
+		return false, err
 	}
+	return true, nil
+}
 
-	return nil
+// setupDockerForwardingPolicy creates a policy to allow forwarding to anywhere to the docker
+// zone (where packets will be dealt with by docker's usual/non-firewalld configuration).
+// The bool return value is true if a firewalld reload is required.
+func setupDockerForwardingPolicy() (bool, error) {
+	// https://firewalld.org/documentation/man-pages/firewalld.dbus.html#FirewallD1.config
+	policy := map[string]interface{}{
+		"version":       "1.0",
+		"description":   "allow forwarding to the docker zone",
+		"ingress_zones": []string{"ANY"},
+		"egress_zones":  []string{dockerZone},
+		"target":        "ACCEPT",
+	}
+	if err := connection.sysConfObj.Call(dbusInterface+".config.addPolicy", 0, dockerFwdPolicy, policy).Err; err != nil {
+		var derr dbus.Error
+		if errors.As(err, &derr) {
+			if derr.Name == "org.fedoraproject.FirewallD1.Exception" && strings.HasPrefix(err.Error(), "NAME_CONFLICT") {
+				log.G(context.TODO()).Debugf("Firewalld: %s policy already exists", dockerFwdPolicy)
+				return false, nil
+			}
```

Comment:
```suggestion
			if derr.Name == dbusInterface+".Exception" && len(derr.Args) > 0 {
				if msg, ok := derr.Args[0].(string); ok && strings.HasPrefix(msg, "NAME_CONFLICT") {
					log.G(context.TODO()).Debugf("Firewalld: %s policy already exists", dockerFwdPolicy)
					return false, nil
				}
			}
```

Probably overkill to unpack the error args, but I was curious how awkward it would be to type-assert. Honestly, it's not too awful with the length check lifted into the outer condition.

Reply:
Why, when dbus's Error.Error() is already doing it? ...

https://github.com/moby/moby/blob/faf84d7f0a1f2e6badff6f720a3e1e559c356fff/vendor/github.com/godbus/dbus/v5/conn.go#L704-L712

Reply:
Ahh, you know something I didn't.

Reply:
```suggestion
			if derr.Name == dbusInterface+".Exception" && strings.HasPrefix(err.Error(), "NAME_CONFLICT") {
				log.G(context.TODO()).Debugf("Firewalld: %s policy already exists", dockerFwdPolicy)
				return false, nil
			}
```